### DISCUSSION TITLE
How to initialise the database connection only when needed?

### DISCUSSION THREAD

**Discussion by waldoc89** [fastapi/fastapi#4223] [Questions] [answered]
I am building a simple API. Most of the requests do not require any database connections. 
My intention is to prevent any useless database pings/sessions made by the clients. 

Following the documentation on FastAPI website - 
https://fastapi.tiangolo.com/tutorial/sql-databases/?h=database

I'm not sure how to initialize the database session only when needed, in situations where a router URL request *does* use the DB session instance. I understand (please correct me if I'm wrong) that the function `def get_db()` set at the `main.py` file is the one that does that. 
Is it a common practice to use it in router files under specific circumstances? Is there an implementation example for this somewhere?

**Comment by simondale00** [comment, marked as answer]
Hey @waldoc89.

I would recommend the patterns recommended in the tutorial you've referenced. This pattern only initializes a connection to the database _when a query is issued_, meaning there are no redundant database pings/sessions made by the clients, at least when using SQLAlchemy. I'll explain why by referencing various pieces of the code from the tutorial...

*database.py*
```
engine = create_engine(
    SQLALCHEMY_DATABASE_URL, connect_args={"check_same_thread": False}
)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
```

^ This snippet is from the ["Create a Base class"](https://fastapi.tiangolo.com/tutorial/sql-databases/?h=database#create-a-base-class) section. This creates a global `engine` variable, which amongst other things, maintains a connection pool to your database. However, by default, creation of an `engine` does not immediately connect to your backend database.

The second variable, `SessionLocal`, is a session factory. This line simply configures how you want to create future sessions; it does not create a session, create a connection, or issue any queries.


*main.py*
```
# Dependency
def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()
        
@app.post("/users/", response_model=schemas.User)
def create_user(user: schemas.UserCreate, db: Session = Depends(get_db)):
    db_user = crud.get_user_by_email(db, email=user.email)
    if db_user:
        raise HTTPException(status_code=400, detail="Email already registered")
    return crud.create_user(db=db, user=user)
```

^ This snippet is from the ["Create a dependency"](https://fastapi.tiangolo.com/tutorial/sql-databases/?h=database#create-a-dependency) section.

The `get_db` method is designed to create a session (using our `SessionLocal` factory) and yield it to the caller. The session is cleaned up when the caller is finished. _However_, SQLAlchemy will not connect to your database or issue a query by when the session is created. Default behaviour is to initiate the connection when you first issue a query ([reference](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#what-does-the-session-do)).

Moreover, you can invoke `get_db` as a dependency only by those endpoints that use the database. You can see an example in the `create_user` signature; The keyword argument `db: Session = Depends(get_db)` tells FastAPI to invoke your `get_db` method before calling the path operation.

Best,
Simon
